systemd

lorenzleutgeb opened the pull request:
Progress towards #82. As of 2025-10-16, this is only a draft, to get early feedback.

Please read the commit messages. I realized that the two implementations are actually leaked through the respective traits. This prevents cleanly providing a third implementation. Before I go any further, I'd like your feedback. Because honestly if we don't figure this out and agree, it would not make much sense to continue.

windy1 commented:
There seems to be a little confusion, so maybe I can help clear this up before I provide a more granular review. The purpose of this library is to provide a cross-platform interface for platform-specific mDNS implementations, i.e. Avahi and Bonjour. Therefore, there is no need to add a feature flag for the individual libraries because Bonjour is the de facto standard for macOS/Windows, as Avahi is for Linux. If we were to add additional implementations per-platform, then a feature would be appropriate, but as it stands now, you must have at least one of these.

To implement your requested feature, my recommendation would be to allow registration of some kind of side effect handler/callback that can be invoked in the implementations. What you've previously described is something core to how mDNS works, so both Avahi and Bonjour should be implemented on the common traits. That said, I am not inherently against providing implementation specific interfaces should there be a specific behavior that is present in one implementation, but not another that we wish to expose, but that does not seem to be the case here.

Hope this helps you get a better understanding of the motive of this library, and I am happy to answer any questions you may have!

lorenzleutgeb commented:
> One caveat, I would like these changes to be opt-in. I can see you have already made avahi the default, but thought it would be worth mentioning.

### Feature Flag for Avahi

I have added the feature flag `"avahi"` for supporting Avahi, but at the same time also added it to the list of default features. This means that all dependents that did not explicitly select `default-features = false`, will not notice a change. Those dependents that *have* selected `default-features = false` in their `Cargo.toml`, will have to make a very simple change, like so:
```diff
 [dependencies]
-zeroconf = { version = "...", default-features = false, features = [ "serde" ] } example
+zeroconf = { version = "...", default-features = false, features = [ "serde", "avahi" ] }
```
(The `"serde"` feature here is just an example. Their list of enabled features might be empty.)

### Feature Flag for systemd

All changes (to Rust code in this crate, and to the declaration of its dependencies in `Cargo.toml`) that relate to `systemd-resolved` would be guarded by a feature flag `"systemd"`.

### Additivity of Features

To be compatible with [Cargo feature unification](https://doc.rust-lang.org/cargo/reference/features.html#feature-unification) all features must be additive. That means, the crate must compile if dependents decide to enable both `"avahi"` and `"systemd"`. Combine this with the desire for backwards compatibility, and this means that the types `zeroconf::{MdnsBrowser,MdnsService}` must always map to the Avahi implementations.

### Submodule for systemd

Just like the module `zeroconf::avahi` would be guarded by the feature flag `"avahi"`, the module `zeroconf::systemd` I am suggesting to introduce would be guarded by the feature flag `"systemd"`.

To help dependents choose at runtime which of the two implementations they might want to use, I suggest to implement a `fn zeroconf::detect_implementation() -> zeroconf::Implementation` or similar, where

```rust
enum Implementation {
  #[cfg(all(feature = "avahi", target_os = "linux")]
  Avahi,
  #[cfg(any(target_vendor = "apple", target_vendor = "pc"))]
  Bonjour,
  #[cfg(all(feature = "systemd", target_os = "linux")]
  Systemd,
}
```

Note at this point, that this is *all optional*, and no dependent would be forced to call `detect_implementation` just by the fact that systemd-resolved would be supported. The exported types would still remain where they are, and dependents that want to use systemd would *opt-in* as you asked.

### Leaky Abstraction
However to get there, some types must be associated to the trait, and not leaked to an exported name of the crate and used in the trait definitions. The leak must have happened because the authors did not expect another implementation to pop up. See f1c21f618f847571d9bf74d6042006795eb563cd
